Make bootstrap client cert loading part of rotation

smarterclayton opened the pull request:
Ensure that bootstrap+client-cert rotation in the Kubelet can:

1. happen in the background so that static pods are launched immediately
2. collapse down to a single CSR request codepath
3. reorganize the code to allow future improvements to fetching bootstrap creds

Reorganize how the kubelet client config is determined. If rotation is
off, simplify the code path. If rotation is on, load the config
from disk, and then pass that into the cert manager. The cert manager
creates a client each time it tries to request a new cert.

Preserves existing behavior where:

1. bootstrap kubeconfig is used if the current kubeconfig is invalid/expired
2. we create the kubeconfig file based on the bootstrap kubeconfig, pointing to
   the location that new client certs will be placed
3. the newest client cert is used once it has been loaded

reverted in https://github.com/kubernetes/kubernetes/pull/71173, original release note was:
```
When a kubelet is using --bootstrap-kubeconfig and certificate rotation, it no longer waits for bootstrap to succeed before launching static pods.
```

```release-note
NONE
```

Fixes #68686

smarterclayton committed:
Make bootstrap client cert loading part of rotation

Ensure that bootstrap+clientcert-rotation in the Kubelet can:

1. happen in the background so that static pods aren't blocked by bootstrap
2. collapse down to a single call path for requesting a CSR
3. reorganize the code to allow future flexibility in retrieving bootstrap creds

Fetching the first certificate and later certificates when the kubelet
is using client rotation and bootstrapping should share the same code
path. We also want to start the Kubelet static pod loop before
bootstrapping completes. Finally, we want to take an incremental step
towards improving how the bootstrap credentials are loaded from disk
(potentially allowing for a CLI call to get credentials, or a remote
plugin that better integrates with cloud providers or KSMs).

Reorganize how the kubelet client config is determined. If rotation is
off, simplify the code path. If rotation is on, load the config
from disk, and then pass that into the cert manager. The cert manager
creates a client each time it tries to request a new cert.

Preserve existing behavior where:

1. bootstrap kubeconfig is used if the current kubeconfig is invalid/expired
2. we create the kubeconfig file based on the bootstrap kubeconfig, pointing to
   the location that new client certs will be placed
3. the newest client cert is used once it has been loaded

smarterclayton commented:
/assign @awly 

@mikedanese @liggitt as discussed, @aaronlevy will allow bootkube and other self-hosting mechanisms to rely on bootstrapped master nodes and use static pods.

smarterclayton commented:
I reverted #66056 in this - perhaps we could instead do a more aggressive dial timeout for bootstrap credentials (now that we always do both in the background)?

smarterclayton commented:
Testing this here and in openshift/origin#21274 so I have a better baseline across different environments

smarterclayton commented:
There are four core scenarios to test:

1. rotation on, bootstrap on
  1. no client certs, --kubeconfig file doesn't exist -> kubeconfig is created pointed at destination of client certs, cert is refreshed
  2. client certs exist, --kubeconfig file doesn't exist -> kubeconfig is created pointing at client certs, cert is not refreshed
  3. client certs exist, --kubeconfig exists, certs are valid -> cert is not refreshed until next interval
  3. client certs exist, --kubeconfig exists, certs are expired -> static pod starts and cert is refreshed
2. rotation off, bootstrap on
  1. bootstrap cert is expired -> rebootstrap
  2. bootstrap file doesn't exist -> exit with error
  3. regular kubeconfig exists but is invalid -> exit with error
3. rotation off, bootstrap off
  1. kubeconfig doesn't exist -> exit with error
  2. kubeconfig exists but is invalid/expired -> static pods start, errors are logged
4. rotation on, bootstrap off

awly commented on cmd/kubelet/app/server.go:
```diff
@@ -771,6 +731,74 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.Dependencies, stopCh <-chan
+			// XXX: When an external bootstrap source is available, it should be possible to always use that source
+			//      to retrieve new credentials.
+			config := certConfig
+			if current != nil {
```
IIUC, if certificate manager has an expired cert and fails to rotate it, `Current()` will return that expired cert.
Can we check that `current` is still valid here and fall back to certConfig if it's not?

smarterclayton replied:
Current() returns nil if the cert is expired.

smarterclayton replied:
Or at least it should, that was the intended behavior a while back.  The transport rotation loop checks for the transition from valid -> invalid and uses that as the "die if this persists"

awly replied:
The cached cert is only updated on successful rotation: http://go/gh/kubernetes/kubernetes/blob/3c493bc95a825f0f6eb38ef78a19811f204053ec/staging/src/k8s.io/client-go/util/certificate/certificate_manager.go#L396
And `Current()` just returns the cached cert without checking expiry http://go/gh/kubernetes/kubernetes/blob/3c493bc95a825f0f6eb38ef78a19811f204053ec/staging/src/k8s.io/client-go/util/certificate/certificate_manager.go#L204-L208

smarterclayton replied:
Ah, I added the filtering to getCurrentCertificateOrBootstrap.  There's very little point in the cert manager handing me back an expired cert - can you think of a case where that makes sense?  If not, we may want to change the signature and verify callers get the right thing (callers should have to deal with nil in general).  Alternatively, we make every caller filter for a valid cert.

smarterclayton replied:
ok, I agree.  I'll make that change and add tests

awly commented on pkg/kubelet/certificate/bootstrap/bootstrap.go:
```diff
@@ -43,6 +38,55 @@ import (
+		return clientConfig, clientConfig, nil
+	}
+
+	store, err := certificate.NewFileStore("kubelet-client", certDir, certDir, "", "")
```
Is this only used to generate `pemPath`?
moveit down to where it's used.

smarterclayton replied:
it's high in the function to fail earlier on local config

awly commented on pkg/kubelet/certificate/bootstrap/bootstrap.go:
```diff
@@ -43,6 +38,55 @@ import (
+// kubeconfigPath on disk is populated based on bootstrapPath but pointing to the location of the client cert
+// in certDir. This preserves the historical behavior of bootstrapping where on subsequent restarts the
+// most recent client cert is used to request new client certs instead of the initial token.
+func LoadClientConfig(kubeconfigPath string, bootstrapPath string, certDir string) (certConfig, userConfig *restclient.Config, err error) {
```
bootstrapPath, certDir string

awly commented on cmd/kubelet/app/server.go:
```diff
@@ -771,6 +731,74 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.Dependencies, stopCh <-chan
+		return &transportConfig, closeAllConns, nil
+	}
+
+	if len(s.BootstrapKubeconfig) > 0 {
```
Should this get called before starting the cert managed above?

smarterclayton replied:
this is in the unmanaged path.  In the managed path bootstrap is managed by the manager.

awly reviewed (commented):
Re #66056: it should be fine to remove if we add aggressive timeout and _fast_ retries for initial bootstrap.

For context: it was needed on GCE because network programming can lag behind and bootstrapping would get trapped in a very long timeout, even after master IP becomes reachable from Node. cc @mikedanese

fedebongio commented:
/assign @caesarxuchao

smarterclayton commented:
/assign @awly

smarterclayton commented:
still looking at making the waitForServer path able to have a more aggressive timeout and dial.

smarterclayton commented:
Ok, I *think* the last commit addresses the issue, but still need to test.  Using a custom dial function on a rest client config busts the cache, so we always get a new transport.  I refactored wait (long time coming anyway) to give us sequence numbers.  We gradually increase the time we wait until we hit a limit, only retrying on timeout.  Feels a bit awkward to treat timeout special, but is close enough for government work.

Only problem with the commit as is that it fills the TLS cache with infinite entries if anyone used LoadClientCert too much.  So I'll change this before we merge to bypass the TLS cache completely, or talk with @liggitt about whether we need to make the cache an LRU or bypass the cache with a custom dial function.

liggitt commented:
> Only problem with the commit as is that it fills the TLS cache with infinite entries if anyone used LoadClientCert too much. So I'll change this before we merge to bypass the TLS cache completely, or talk with @liggitt about whether we need to make the cache an LRU or bypass the cache with a custom dial function.

Custom dial functions must produce unique cache keys. 

If you want to guarantee no stuck tcp connections after timeout, I think having an option to disable keep alive when constructing the transport would be better

smarterclayton commented:
/retest

smarterclayton commented:
/retest

mikedanese commented on pkg/kubelet/certificate/bootstrap/bootstrap.go:
```diff
@@ -98,23 +139,47 @@ func LoadClientCert(kubeconfigPath string, bootstrapPath string, certDir string,
 	}
+	for interval.Steps > 0 {
+		// creating a new client with a new Dial function ensures we get a new cache entry
+		// TODO: it also leaks cache entries, which is a broader problem
```
https://github.com/kubernetes/kubernetes/pull/58677 would help here.

We need to make sure transports are also cleaned up. I don't think that will happen properly when bootstrapClient is GC'd:

https://github.com/golang/go/issues/24739

smarterclayton replied:
It won't, which is why Jordan suggested a "bypass reuse" option.  I don't think the TLS cache holding things forever is a reasonable long term path, but I could be persuaded it's a short term avoidable problem.

I could just create a separate transport here and bypass the cache completely, which may just be the simplest option.

awly commented on pkg/kubelet/certificate/bootstrap/bootstrap.go:
```diff
@@ -98,23 +140,64 @@ func LoadClientCert(kubeconfigPath string, bootstrapPath string, certDir string,
+
+		bootstrapClient, err := certificates.NewForConfig(timeoutClientConfig)
+		if err != nil {
+			if t, ok := err.(timeoutError); ok && t.Timeout() {
```
Why would creating a new certs client timeout?

smarterclayton replied:
This is the first api call made to join a node to a cluster, it races with network programming on GCE, GCE has bugs where ACKs disappear into the ether.  See the referenced PR

awly replied:
But isn't `csr.RequestNodeCertificate` below the API call? Why would `certificates.NewForConfig` result in a request?

smarterclayton replied:
Moved out of this PR so as not to block this if we can't get review time by tomorrow.

awly commented on staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:
```diff
@@ -177,6 +177,32 @@ type Backoff struct {
+}
+
+// Step returns the next interval in the exponential backoff.
+func (b *Backoff) Step() time.Duration {
```
I know it's annoying, but can you move this refactor to separate PR?

smarterclayton replied:
I will once we tease apart the retry design.

smarterclayton commented on pkg/kubelet/certificate/bootstrap/bootstrap.go:
```diff
@@ -98,23 +140,64 @@ func LoadClientCert(kubeconfigPath string, bootstrapPath string, certDir string,
+		timeoutClientConfig.Transport = &http.Transport{
+			TLSClientConfig:   tlsConfig,
+			DisableKeepAlives: true,
+			DialContext:       (&net.Dialer{Timeout: t}).DialContext,
```
No, it’s not thread safe to change a dialect under the covers.  Default dialers aren’t expensive.

smarterclayton commented on pkg/kubelet/certificate/bootstrap/bootstrap.go:
```diff
@@ -43,11 +41,60 @@ import (
+		return nil, nil, err
+	}
+
+	// use the current client config
```
Should we?  Probably.  But this is the current behavior.  And after five minutes without a rotation the kubelet restarts.  So preserving legacy behavior.

mikedanese commented on pkg/kubelet/certificate/bootstrap/bootstrap.go:
```diff
@@ -98,23 +140,64 @@ func LoadClientCert(kubeconfigPath string, bootstrapPath string, certDir string,
+		t := interval.Step()
+		timeoutClientConfig.Transport = &http.Transport{
+			TLSClientConfig:   tlsConfig,
+			DisableKeepAlives: true,
```
See here https://github.com/kubernetes/kubernetes/pull/69890#discussion_r232780288 for justification. Maybe add a comment?

smarterclayton committed:
Ensure the bootstrap rotation code is tested by forcing rotation

Expose both a Stop() method (for cleanup) and a method to force
cert rotation, but only expose Stop() on the interface.

Verify that we choose the correct client.

smarterclayton commented:
Add a test to verify we don't regress the core client behavior (use the bootstrap client, then the retrieved client).

smarterclayton commented:
/retest

smarterclayton commented:
@awly this is back to effectively the original PR you reviewed.  Was your previous review sufficient or were you looking for more?

mikedanese reviewed (commented):
I'm +1 on getting this in to CI and seeing how it goes.

/lgtm
/approve

liggitt commented on cmd/kubelet/app/server.go:
```diff
@@ -610,28 +584,15 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.Dependencies, stopCh <-chan
 		}
 
-		// csiClient works with CRDs that support json only
-		clientConfig.ContentType = "application/json"
```
pre-merge tests wouldn't pick up failures here, since it's alpha. need to verify this still works before merge

smarterclayton replied:
I'll restore with comments linking to the issue.  I'm surprised at renegotiation not working on watch, I thought we had tests for it.

liggitt commented:
/hold
for the csiClient question

smarterclayton commented on cmd/kubelet/app/server.go:
```diff
@@ -771,6 +732,81 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.Dependencies, stopCh <-chan
+	newClientFn := func(current *tls.Certificate) (certificatesclient.CertificateSigningRequestInterface, error) {
+		// If we have a valid certificate, use that to fetch CSRs. Otherwise use the bootstrap
+		// credentials.
+		// XXX: When an external bootstrap source is available, it should be possible to always use that source
```
picked up another team's convention and used it here.

smarterclayton commented on cmd/kubelet/app/server.go:
```diff
@@ -537,66 +538,39 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.Dependencies, stopCh <-chan
 		kubeDeps.EventClient = nil
 		kubeDeps.HeartbeatClient = nil
 		klog.Warningf("standalone mode, no API client")
+
```
intentional to make the blocks easier to read.  I may make this a switch instead.

smarterclayton commented:
Comments addressed

liggitt commented:
/hold cancel

smarterclayton commented:
/retest

smarterclayton commented:
This was reverted in #71173 due to breaking kubeadm's requirement that `--kubeconfig` be provided on masters as a high powered credential and cause a rotation of the cert.  Restored in #71174 with coverage for kubeadm's flow.
